src/core/net/ip4_headers.cpp

```diff
@@ -0,0 +1,59 @@
+/*
+ *  Copyright (c) 2022, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements IP4 headers processing.
+ */
+
+#include "ip4_headers.hpp"
+
+namespace ot {
+namespace Ip4 {
+
+Error Header::ParseFrom(const Message &aMessage)
+{
+    Error error = kErrorParse;
+
+    SuccessOrExit(error = aMessage.Read(0, *this));
+    VerifyOrExit(IsValid());
+    VerifyOrExit(GetTotalLength() == aMessage.GetLength());
+
+    error = kErrorNone;
```

Comment:
The logic here does not seem to be correct. We will return `kErrorNone` independent of `IsValid()` and `GetTotalLength()` checks since `error` would be set to `kErrorNone` from `SuccessOrExit(error = aMessage.Read(0, *this));` line.

I guess you copied this from similar `Ip6::Header` method. Note that we do not set the `error` in `Ip6::Header::ParseFrom()` implementation in `SuccessOrExit(aMessage.Read(0, *this));`:

```c++
Error Header::ParseFrom(const Message &aMessage)
{
    Error error = kErrorParse;

    SuccessOrExit(aMessage.Read(0, *this));
    VerifyOrExit(IsValid());
    VerifyOrExit(sizeof(Header) + GetPayloadLength() == aMessage.GetLength());

    error = kErrorNone;

exit:
    return error;
}
```

Reply:
Yes, I added the `error=` by mistake, thanks.